sdk/python/lib/pulumi/provider/server.py

```diff
@@ -0,0 +1,231 @@
+    `Construct` and other methods invoked by the engine when the user
+    program creates a remote `ComponentResource` (with `remote=true`
+    in the constructor).
+
+    See `ResourceProvider` defined in `provider.proto`.
+
+    """
+
+    engine_address: str
+    provider: Provider
+    args: List[str]
+
+    # NOTE: remove @_asynchronized when we can avoid modifying globals in the method body.
+    @_asynchronized
+    async def Construct(self, request: proto.ConstructRequest, context) -> proto.ConstructResponse:  # pylint: disable=invalid-overridden-method
+        assert isinstance(request, proto.ConstructRequest), \
+            f'request is not ConstructRequest but is {type(request)} instead'
+
+        # Should we set these settings back after we are done?
+
+        pulumi.runtime.settings.reset_options(
+            project=_empty_as_none(request.project),
+            stack=_empty_as_none(request.stack),
+            parallel=_zero_as_none(request.parallel),
+            engine_address=self.engine_address,
+            monitor_address=_empty_as_none(request.monitorEndpoint),
+            preview=request.dryRun)
+
+        pulumi.runtime.config.set_all_config(dict(request.config))
+
+        inputs = self._construct_inputs(request)
+
+        result = self.provider.construct(name=request.name,
+                                         resource_type=request.type,
+                                         inputs=inputs,
+                                         options=self._construct_options(request))
+
+
+        response = await self._construct_response(result)
+
+        # Wait for outstanding RPCs such as more provider Construct
+        # calls. This can happen if i.e. provider creates child
+        # resources but does not await their URN promises.
+        #
+        # Do not await all tasks as that starts hanging waiting for
+        # indefinite grpc.aio servier tasks.
+        await wait_for_rpcs(await_all_outstanding_tasks=False)
+
+        return response
+
+    @staticmethod
+    def _construct_inputs(request: proto.ConstructRequest) -> Dict[str, pulumi.Output]:
+
+        def deps(key: str) -> Set[Resource]:
+            return set(DependencyResource(urn) for urn in
+                       request.inputDependencies.get(
+                           key,
+                           proto.ConstructRequest.PropertyDependencies()
+                       ).urns)
+
+        return {
+            k: ProviderServicer._construct_output(the_input, deps=deps(k))
+            for k, the_input in
+            rpc.deserialize_properties(request.inputs, keep_unknowns=True).items()
+        }
+
+    @staticmethod
+    def _construct_output(the_input: Any, deps: Set[Resource]) -> Any:
+        is_secret = rpc.is_rpc_secret(the_input)
+
+        # If it's a prompt value, return it directly without wrapping
+        # it as an output.
+        if not is_secret and len(deps) == 0:
+            return the_input
```

Comment:
We have a basic integration test for prompt values at https://github.com/pulumi/pulumi/tree/master/tests/integration/construct_component_plain.

I could imagine implementing a `testcomponent-python` similar to the integration test in this PR.